src/test/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandlerTests.java

```diff
@@ -264,4 +281,94 @@ public void testOutboundWithAsyncResults() {
+		ProducerFactory<Integer, String> producerFactory = new DefaultKafkaProducerFactory<>(
+				KafkaTestUtils.producerProps(embeddedKafka));
+		ReplyingKafkaTemplate<Integer, String, String> template = new ReplyingKafkaTemplate<>(producerFactory, container);
+		template.start();
+		assertThat(assigned.await(30, TimeUnit.SECONDS)).isTrue();
+		KafkaProducerMessageHandler<Integer, String> handler = new KafkaProducerMessageHandler<>(template);
+		handler.setBeanFactory(mock(BeanFactory.class));
+		QueueChannel replies = new QueueChannel();
+		handler.setOutputChannel(replies);
+		handler.afterPropertiesSet();
+
+		Message<?> message = MessageBuilder.withPayload("foo")
+				.setHeader(KafkaHeaders.TOPIC, topic5)
+				.setHeader(KafkaHeaders.MESSAGE_KEY, 2)
+				.setHeader(KafkaHeaders.PARTITION_ID, 1)
+				.build();
+		handler.handleMessage(message);
+
+		ConsumerRecord<Integer, String> record = KafkaTestUtils.getSingleRecord(consumer, topic5);
+		assertThat(record).has(key(2));
+		assertThat(record).has(partition(1));
+		assertThat(record).has(value("foo"));
+		Map<String, Object> headers = new HashMap<>();
+		new DefaultKafkaHeaderMapper().toHeaders(record.headers(), headers);
+		assertThat(headers.get(KafkaHeaders.REPLY_TOPIC)).isEqualTo(topic6.getBytes());
+		ProducerRecord<Integer, String> pr = new ProducerRecord<>(topic6, 0, 1, "FOO", record.headers());
+		template.send(pr);
+		Message<?> reply = replies.receive(30_000);
+		assertThat(reply).isNotNull();
+		assertThat(reply.getPayload()).isEqualTo("FOO");
+		assertThat(reply.getHeaders().get(KafkaHeaders.TOPIC)).isNull();
+		assertThat(reply.getHeaders().get(KafkaHeaders.CORRELATION_ID)).isNull();
+
+		message = MessageBuilder.withPayload("foo")
+				.setHeader(KafkaHeaders.TOPIC, topic5)
+				.setHeader(KafkaHeaders.MESSAGE_KEY, 2)
+				.setHeader(KafkaHeaders.PARTITION_ID, 1)
+				.setHeader(KafkaHeaders.REPLY_TOPIC, "bad")
+				.build();
+		try {
+			handler.handleMessage(message);
+			fail("Expected exception");
+		}
+		catch (MessageHandlingException e) {
+			assertThat(e.getCause().getMessage())
+					.isEqualTo("The reply topic header [bad] does not match any reply container topic: "
+							+ "[" + topic6 + "]");
+		}
+		message = MessageBuilder.withPayload("foo")
+				.setHeader(KafkaHeaders.TOPIC, topic5)
+				.setHeader(KafkaHeaders.MESSAGE_KEY, 2)
+				.setHeader(KafkaHeaders.PARTITION_ID, 1)
+				.setHeader(KafkaHeaders.REPLY_PARTITION, 999)
+				.build();
+		try {
+			handler.handleMessage(message);
+			fail("Expected exception");
+		}
+		catch (MessageHandlingException e) {
+			assertThat(e.getCause().getMessage())
+					.isEqualTo(
+							"The reply partition header [999] does not match any reply container partition for topic ["
+									+ topic6 + "]: [0, 1]");
+		}
+
+		template.stop();
+		// discard from the test consumer
+		KafkaTestUtils.getSingleRecord(consumer, topic6);
```

Comment:
Uh! Looks like we are missing to call `consumer.close()` in this test class. I guess the `@AfterClass` is the best place.